src/realm/object-store/impl/deep_change_checker.cpp

```diff
@@ -0,0 +1,345 @@
+        for (auto key_path_array : m_key_path_arrays) {
+            for (auto key_path : key_path_array) {
+                if (key_path.size() != 0) {
+                    m_filtered_columns_in_root_table.push_back(key_path[0].second);
+                }
+                for (auto key_path_element : key_path) {
+                    m_filtered_columns.push_back(key_path_element.second);
+                }
+            }
+        }
+    }
+}
+
+bool DeepChangeChecker::check_outgoing_links(Table const& table, int64_t object_key,
+                                             std::vector<ColKey> filtered_columns, size_t depth)
+{
+    auto table_key = table.get_key();
+
+    // First we create an iterator pointing at the table identified by `table_key` within the `m_related_tables`.
+    auto it = find_if(begin(m_related_tables), end(m_related_tables), [&](auto&& tbl) {
+        return tbl.table_key == table_key;
+    });
+    // If no iterator could be found the table is not contained in `m_related_tables` and we cannot check any
+    // outgoing links.
+    if (it == m_related_tables.end())
+        return false;
+    // Likewise if the table could be found but does not have any (outgoing) links.
+    if (it->links.empty())
+        return false;
+
+    // Check if we're already checking if the destination of the link is
+    // modified, and if not add it to the stack
+    auto already_checking = [&](int64_t col) {
+        auto end = m_current_path.begin() + depth;
+        auto match = std::find_if(m_current_path.begin(), end, [&](auto& p) {
+            return p.object_key == object_key && p.col_key == col;
+        });
+        if (match != end) {
+            for (; match < end; ++match)
+                match->depth_exceeded = true;
+            return true;
+        }
+        m_current_path[depth] = {object_key, col, false};
+        return false;
+    };
+
+    const Obj obj = table.get_object(ObjKey(object_key));
+    auto linked_object_changed = [&](OutgoingLink const& link) {
+        if (already_checking(link.col_key))
+            return false;
+
+        if (ColKey(link.col_key).get_type() == col_type_BackLink) {
+            // Related tables can include tables that are only reachable via backlinks.
+            // These tables do not need to be considered when executing this check and
+            // therefore be ignored.
+            return false;
+        }
+
+        if (!link.is_list) {
+            ObjKey dst_key = obj.get<ObjKey>(ColKey(link.col_key));
+            if (!dst_key) // do not descend into a null or unresolved link
+                return false;
+            return check_row(*table.get_link_target(ColKey(link.col_key)), dst_key.value, filtered_columns,
+                             depth + 1);
+        }
+
+        auto& target = *table.get_link_target(ColKey(link.col_key));
+        auto lvr = obj.get_linklist(ColKey(link.col_key));
+        return std::any_of(lvr.begin(), lvr.end(), [&, this](auto key) {
```

Comment:
is capturing `[&]` insufficient to capture `this` as well? The `this` here is a bit confusing because you're already inside a lambda capturing `[&]` so `this` is pointing to the class that will be generated for the lambda and not the `this` of `DeepChangeChecker`, but if the compiler couldn't figure all that out I guess this is okay.

Also, `const auto& key` please